src/wallet/wallet_tx_builder.cpp

```diff
@@ -474,83 +639,10 @@ WalletTxBuilder::ResolveInputsAndPayments(
-                if (canResolveOrchard
-                    && ua.GetOrchardReceiver().has_value()
-                    && (strategy.AllowRevealedAmounts() || payment.GetAmount() <= maxOrchardAvailable)
-                    ) {
-                    resolvedPayments.emplace_back(
-                        ua, ua.GetOrchardReceiver().value(), payment.GetAmount(), payment.GetMemo(), false);
-                    if (!strategy.AllowRevealedAmounts()) {
-                        maxOrchardAvailable -= payment.GetAmount();
-                    }
-                    orchardOutputs += 1;
-                } else if (ua.GetSaplingReceiver().has_value()
-                    && (strategy.AllowRevealedAmounts() || payment.GetAmount() <= maxSaplingAvailable)
-                    ) {
-                    resolvedPayments.emplace_back(
-                        ua, ua.GetSaplingReceiver().value(), payment.GetAmount(), payment.GetMemo(), false);
-                    if (!strategy.AllowRevealedAmounts()) {
-                        maxSaplingAvailable -= payment.GetAmount();
-                    }
-                } else {
-                    if (strategy.AllowRevealedRecipients()) {
-                        if (ua.GetP2SHReceiver().has_value()) {
-                            resolvedPayments.emplace_back(
-                                ua, ua.GetP2SHReceiver().value(), payment.GetAmount(), std::nullopt, false);
-                        } else if (ua.GetP2PKHReceiver().has_value()) {
-                            resolvedPayments.emplace_back(
-                                ua, ua.GetP2PKHReceiver().value(), payment.GetAmount(), std::nullopt, false);
-                        } else {
-                            // This should only occur when we have
-                            // • an Orchard-only UA,
-                            // • `AllowRevealedRecipients`, and
-                            // • can’t resolve Orchard (which means either insufficient non-Sprout
-                            //   funds or pre-NU5).
-                            resolutionError = AddressResolutionError::CouldNotResolveReceiver;
-                        }
-                    } else if (strategy.AllowRevealedAmounts()) {
-                        resolutionError = AddressResolutionError::TransparentReceiverNotAllowed;
-                    } else {
-                        resolutionError = AddressResolutionError::RevealingReceiverAmountsNotAllowed;
-                    }
-                }
-            }
-        });
-
-        if (resolutionError.has_value()) {
-            return tl::make_unexpected(resolutionError.value());
+        auto res = ResolvePayment(payment, canResolveOrchard, strategy, maxSaplingAvailable, maxOrchardAvailable, orchardOutputs);
+        res.map([&](const ResolvedPayment& rpayment) { resolvedPayments.emplace_back(rpayment); });
+        if (!res.has_value()) {
+            return tl::make_unexpected(res.error());
```

Comment:
This isn't `return res` because of the need to embed the `AddressResolutionError` error type in the larger `InputSelectionError` type.